tests/model/document/document.js

```diff
@@ -276,49 +275,160 @@ describe( 'Document', () => {
-
-			const position = new Position( root, [ 2 ] );
-			const selectionPosition = doc.getNearestSelectionPosition( position );
+			doc.schema.registerItem( 'image' );
+			doc.schema.allow( { name: 'image', inside: '$root' } );
+			doc.schema.registerItem( 'widget', '$block' );
+			doc.schema.allow( { name: 'widget', inside: '$root' } );
+			doc.schema.objects.add( 'widget' );
 
-			expect( selectionPosition.path ).to.deep.equal( [ 2, 0 ] );
+			root = doc.createRoot();
+			selection = doc.selection;
 		} );
 
-		it( 'should return first position in root if there is no position where text node can be placed', () => {
-			root.appendChildren( new Element( 'table' ) );
-
-			const position = new Position( root, [ 0, 0 ] );
-			const selectionPosition = doc.getNearestSelectionPosition( position );
-
-			expect( selectionPosition.path ).to.deep.equal( [ 0 ] );
-		} );
+		test(
+			'should return collapsed range if text node can be placed at that position - both',
+			'<paragraph>[]</paragraph>',
+			'both',
+			'<paragraph>[]</paragraph>'
+		);
+
+		test(
+			'should return collapsed range if text node can be placed at that position - forward',
+			'<paragraph>[]</paragraph>',
+			'forward',
+			'<paragraph>[]</paragraph>'
+		);
+
+		test(
+			'should return collapsed range if text node can be placed at that position - backward',
+			'<paragraph>[]</paragraph>',
+			'backward',
+			'<paragraph>[]</paragraph>'
+		);
+
+		test( 'should return null in empty document - both', '', 'both', null );
+
+		test( 'should return null in empty document - backward', '', 'backward', null );
+
+		test( 'should return null in empty document - forward', '', 'forward', null );
+
+		test(
+			'should find range before when searching both ways',
+			'<paragraph></paragraph>[]<paragraph></paragraph>',
+			'both',
+			'<paragraph>[]</paragraph><paragraph></paragraph>'
+		);
+
+		test(
+			'should find range before when searching backward',
+			'<paragraph></paragraph>[]<paragraph></paragraph>',
+			'backward',
+			'<paragraph>[]</paragraph><paragraph></paragraph>'
+		);
+
+		test(
+			'should find range after when searching forward',
+			'<paragraph></paragraph>[]<paragraph></paragraph>',
+			'forward',
+			'<paragraph></paragraph><paragraph>[]</paragraph>'
+		);
+
+		test(
+			'should find range after when searching both ways when it is closer',
+			'<paragraph></paragraph><image></image>[]<paragraph></paragraph>',
+			'both',
+			'<paragraph></paragraph><image></image><paragraph>[]</paragraph>'
+		);
+
+		test(
+			'should find range before when searching both ways when it is closer',
+			'<paragraph></paragraph><image></image>[]<image></image><image></image><paragraph></paragraph>',
+			'both',
+			'<paragraph>[]</paragraph><image></image><image></image><image></image><paragraph></paragraph>'
+		);
+
+		test(
+			'should return null if there is no valid range',
+			'<image></image>',
```

Comment:
OK... Image is not an object in this test suite. This is confusing. Could we name it like... damn, I can't find a name, because this case shouldn't really occur :D. Perhaps we can call it `<emptyBlock>` then?